Tutorials
=======================================

Using Metadensity in Jupyter notebooks
--------------------------------------
.. toctree::
   :maxdepth: 1

   0_Example_on_test_data.ipynb

More data means more fun
--------------------------------------
from here on, to reproduce the code in the notebooks, you need to download eCLIP data from the ENCODE project, create your dataloader and menifest!

`ENCODE <https://www.encodeproject.org/>`_ 



Starting from skipper data
--------------------------------------

.. toctree::
   :maxdepth: 1

   1_Example_using_skipper_outputs

Create customized features: intronic polyadenylation, miRNA hosting genes
----------------------------------------------------------------------------
.. toctree::
   :maxdepth: 1
   
   0_customize_feature_example_intronic_polyA
   0_customize_feature_example_miRNA_host_gene

Comparing 2 groups of transcripts
--------------------------------------
.. toctree::
   :maxdepth: 1
   
   deseq_metaden

Loading precomputed data
--------------------------------------
.. toctree::
   :maxdepth: 1

   Heatmap_from_precomputed_data

How to build the data needed for other species
-----------------------------------------------
.. toctree::
    :maxdepth: 1

    parse_gencode_coords_into_data

Generalizing to other sequencing technology
-----------------------------------------------
.. toctree::
    :maxdepth: 1

   1_Example_on_Par-CLIP-YBX1.ipynb

Comparing using CIMs and CITs
-----------------------------------------------
.. toctree::
    :maxdepth: 1

    1_Example_on_SF3B4-compare-CIMS-CITS.ipynb

Comparing data from different cell lines
-----------------------------------------------
.. toctree::
    :maxdepth: 1

    1_Example_on_HNRNPC_on_different_cell_line.ipynb

Showing replicates seperately
-----------------------------------------------
.. toctree::
    :maxdepth: 1

    1_Example_on_SF3B4-rep_seperate.ipynb